### ISSUE TITLE
Line numbers cause frame time to increase drastically

### ISSUE THREAD

**Issue by oliverlars** [4coder-editor/4coder#207]
Version: 4.1.4

**Repro Steps (Including a snippet of any potentially related customization code)**
Default 4.1.4 build with line numbers enabled
Open two or more panels with any code file

**Description of the Result**
Frame time increases proportionally to the number of panels open with line numbers.
No longer runs at 60fps with a number of panels > 1

**System**
OS: Windows
Graphics Driver (for graphics bugs): N/A

**Comment by mhalber** [comment]
Hi, just to add to @oliverlars statement - I am also observing significant slowdown with line numbers enabled, especially when opening +10kloc files.

I took a look at frame capture using renderdoc. It seems that 4coder is generating a very large buffer for rendering line-numbers. I tested a file with 25526 lines. The glDrawArrays call for rendering text in the file is taking buffer of size 5160, and takes in the order of 61 microseconds. glDrawArrays for the line numbers suddenly takes buffer of size 765786 and takes in the order of 2250 microseconds.

This still should not cause drop below 60fps, but I suspect 4coder might be spending a lot of time generating the buffer data for rendering line-numbers on the CPU? 

If it is any help, I've attached the renderdoc capture illustrating this issue:
[4coder_line_numbers.zip](https://github.com/4coder-editor/4coder/files/4463629/4coder_line_numbers.zip)

**Comment by aolo2** [comment]
I did some additional markup with builtin profiling tools and found what it was!

For large files almost all the time is spent in `draw_line_number_margin`. More precisely, in this loop:
```
for (;cursor.pos <= visible_range.one_past_last;){
    if (line_number > line_count){
        break;
    }
    ...
    line_number += 1;
}
```

I honestly have no idea what the loop condition does, but the `line_number > line_count` is definitely not doing its job. For large files, especially when viewing the beginning of the buffer, this loop executes thousands of times instead of like 50-100 tops.

I fixed it by getting the line_number from the `visible_range`:
```
one_line_past_last = get_line_number_from_pos(app, buffer, visible_range.one_past_last)
```

And using it as a second break condition:
```
if (line_number > line_count || line_number > one_line_past_last){
    break;
}
```

My repro for this bug was `4coder_base_types.cpp` from the default custom layer. It is fixed with this change.
